### PR TITLE
Adds label filter to qdrant for WEAK_SUPERVISION and MANUAL labels

### PR DESCRIPTION
**Pull request by JWittmeyer**
- https://github.com/code-kern-ai/refinery-submodule-model/pull/53
- https://github.com/code-kern-ai/refinery-neural-search/pull/41
- https://github.com/code-kern-ai/refinery-gateway/pull/153
- https://github.com/code-kern-ai/dev-setup/pull/34

Needs update of qdrant, use dev-setup branch `qdrant-label-filter` for testing
best to test with dev-setup: `bash start -g -b qdrant-label-filter`

Included manual labels need to be valid, which means without conflict or a gold star label. Here, without conflict means that all labels for a labeling task set on one record must be the same for all users to be valid. The payload for the manual label is updated on each label change (add/removal).

The weak supervision labels are uploaded on creation of the embedding and changed on each new weak supervision run.

Further, this PR changes the upload of the payloads to qdrant. Instead of removing and recreating the collection, only the payload is updated. For attribute (filter) changes the whole payload is overwritten, so that also removed attributes are removed from the payload. For changes to the labels, only the `labels` key in the payload is replaced or removed if no labels exist anymore.

The changes affect the following actions:
- weak supervision run or run heuristic + weakly supervise
- add/remove (gold) label
- change filter attribute (add/remove)

Currently no UI integration planned (before refinery 2.0) so only accessible via gates (or GraphiQL):

<details closed>
  <summary>Example GraphiQL request</summary>

```
{
  searchRecordsBySimilarity(
    	projectId: "651d52ea-acc3-4b87-b3e3-dfff018aa7f8", 
    	embeddingId: "1548e4fe-8787-4e2b-956c-d38f32847ebd", 
    	recordId: "2a1408db-ad9f-40f1-abbb-2254b85f97f3", 
    	attFilter: "[{\"key\":\"labels.WEAK_SUPERVISION.clickbait.label\",\"value\":\"yes\"}]") {
    queryLimit
    queryOffset
    fullCount
    sessionId
    recordList {
      recordData
    }
  }
}
```

</details>

<details closed>
  <summary>Example gates request</summary>

```
{
  "headline": "Iraqi vote remains in doubt",
  "@@meta@@":{
    "similarity":{
      "filter":[
        {"key":"labels.WEAK_SUPERVISION.clickbait.label", "value":"no"},
        {"key": "labels.MANUAL.clickbait.label", "value": "no"}
    ]
    }
  }
}
```

</details>

### CODE REVIEW

**Comment by FelixKirschKern** [comment]
reviewed until commit [602427f](https://github.com/code-kern-ai/refinery-neural-search/pull/41/commits/602427feb534faf62b56cb57685aeb19515cba26)
will use this PR for further development

**Review by FelixKirschKern — COMMENTED** [review]
reviewed until commit [602427f](https://github.com/code-kern-ai/refinery-neural-search/pull/41/commits/602427feb534faf62b56cb57685aeb19515cba26)
will use this PR for further development

File: app.py [hunk 1]
```diff
@@ -135,6 +135,42 @@ def create_missing_collections() -> responses.JSONResponse:
...
+class UpdateLabelPayloadsRequest(BaseModel):
+    project_id: str
+    embedding_ids: List[str]
+    record_ids: Optional[List[str]] = None
+
+
+@app.post("/update_label_payloads")
+def update_label_payloads(
+    request: UpdateLabelPayloadsRequest,
+) -> responses.JSONResponse:
```
Comment (JWittmeyer) [inline app.py]: return type

File: app.py [hunk 2]
```diff
@@ -135,6 +135,42 @@ def create_missing_collections() -> responses.JSONResponse:
...
+class UpdateAttributePayloadsRequest(BaseModel):
+    project_id: str
+    embedding_id: str
+    record_ids: Optional[List[str]] = None
+
+
+@app.post("/update_attribute_payloads")
+def update_attribute_payloads(
+    request: UpdateAttributePayloadsRequest,
+) -> responses.JSONResponse:
```
Comment (JWittmeyer) [inline app.py]: return type

File: neural_search/util.py [hunk 3]
```diff
@@ -261,3 +281,113 @@ def detect_outliers(
...
+        record_ids, payloads = zip(*all_object)
+        ids_for_storage = record_ids
+
+    for record_id, payload in zip(record_ids, payloads):
+        if record_id in label_payload_extension:
+            payload["labels"] = label_payload_extension[record_id]
+
+    update_operations = []
+    for point_id, payload in zip(ids_for_storage, payloads):
+        update_operations.append(
```
Comment (JWittmeyer) [inline neural_search/util.py]: this can be done with list comprehension

**Review by FelixKirschKern — APPROVED** [review]

**Pull request merged**